File: Examples/VoidEtching/VoidEtching.cpp
#include <iostream>

#include <lsAdvect.hpp>
#include <lsBooleanOperation.hpp>
#include <lsDomain.hpp>
#include <lsExpand.hpp>
#include <lsMakeGeometry.hpp>
#include <lsPrune.hpp>
#include <lsToSurfaceMesh.hpp>
#include <lsToVoxelMesh.hpp>
#include <lsVTKWriter.hpp>

/**
  3D Example showing how to use the library for topography
  simulation. A substrate with big air inclusions is etched isotropically.
  Voids are detected automatically and only etched once they are exposed.
  \example VoidEtching.cpp
*/

// implement own velocity field
class velocityField : public lsVelocityField<double> {
public:
  double getScalarVelocity(const std::array<double, 3> & /*coordinate*/,
                           int /*material*/,
                           const std::array<double, 3> & /*normalVector*/,
                           unsigned long /*pointId*/) {
    // isotropic etch rate
    return -1;
  }

  std::array<double, 3>
  getVectorVelocity(const std::array<double, 3> & /*coordinate*/,
                    int /*material*/,
                    const std::array<double, 3> & /*normalVector*/,
                    unsigned long /*pointId*/) {
    return std::array<double, 3>({});
  }
};

int main() {

  constexpr int D = 3;
  omp_set_num_threads(4);

  double extent = 30;
  double gridDelta = 1;

  double bounds[2 * D] = {-extent, extent, -extent, extent, -extent, extent};
  lsDomain<double, D>::BoundaryType boundaryCons[D];
  boundaryCons[0] = lsDomain<double, D>::BoundaryType::REFLECTIVE_BOUNDARY;
  boundaryCons[1] = lsDomain<double, D>::BoundaryType::REFLECTIVE_BOUNDARY;
  boundaryCons[2] = lsDomain<double, D>::BoundaryType::INFINITE_BOUNDARY;

  auto substrate =
      lsSmartPointer<lsDomain<double, D>>::New(bounds, boundaryCons, gridDelta);

  double origin[3] = {0., 0., 0.};
  {
    double planeNormal[3] = {0., 0., 1.};
    auto plane = lsSmartPointer<lsPlane<double, D>>::New(origin, planeNormal);
    lsMakeGeometry<double, D>(substrate, plane).apply();
  }

  {
    // create spheres used for booling
    std::cout << "Creating spheres..." << std::endl;
    auto sphere = lsSmartPointer<lsDomain<double, D>>::New(bounds, boundaryCons,
                                                           gridDelta);
    origin[0] = -12;
    origin[1] = -5;
    origin[2] = -15;
    double radius = 10;
    lsMakeGeometry<double, D>(
        sphere, lsSmartPointer<lsSphere<double, D>>::New(origin, radius))
        .apply();
    lsBooleanOperation<double, D> boolOp(
        substrate, sphere, lsBooleanOperationEnum::RELATIVE_COMPLEMENT);
    boolOp.apply();

    origin[0] = -7;
    origin[1] = -30;
    origin[2] = -20;
    radius = 8;
    lsMakeGeometry<double, D>(
        sphere, lsSmartPointer<lsSphere<double, D>>::New(origin, radius))
        .apply();
    // reference to substrate and sphere are kept in boolOp
    boolOp.apply();

    origin[0] = 5;
    origin[1] = 15;
    origin[2] = -2;
    radius = 8;
    lsMakeGeometry<double, D>(
        sphere, lsSmartPointer<lsSphere<double, D>>::New(origin, radius))
        .apply();
    boolOp.apply();

    origin[0] = 2;
    origin[1] = 8;
    origin[2] = -27;
    radius = 8;
    lsMakeGeometry<double, D>(
        sphere, lsSmartPointer<lsSphere<double, D>>::New(origin, radius))
        .apply();
    boolOp.apply();
  }

  // Now etch the substrate isotropically
  auto velocities = lsSmartPointer<velocityField>::New();

  std::cout << "Advecting" << std::endl;

  lsAdvect<double, D> advectionKernel;
  advectionKernel.insertNextLevelSet(substrate);
  advectionKernel.setVelocityField(velocities);
  advectionKernel.setIgnoreVoids(true);

  // Now advect the level set 50 times, outputting every
  // advection step. Save the physical time that
  // passed during the advection.
  double passedTime = 0.;
  unsigned numberOfSteps = 50;
  for (unsigned i = 0; i < numberOfSteps; ++i) {
    std::cout << "\rAdvection step " + std::to_string(i) + " / "
              << numberOfSteps << std::flush;
    auto mesh = lsSmartPointer<lsMesh<>>::New();
    lsToSurfaceMesh<double, D>(substrate, mesh).apply();
    lsVTKWriter<double>(mesh, "void-" + std::to_string(i) + ".vtk").apply();

    advectionKernel.apply();
    passedTime += advectionKernel.getAdvectedTime();
  }
  std::cout << std::endl;

  std::cout << "Time passed during advection: " << passedTime << std::endl;

  return 0;
}

File: examples/VoidEtching/VoidEtching.py
import viennals

viennals.setDimension(3)

# Implement own velocity field
class VelocityField(viennals.VelocityField):
    def getScalarVelocity(self, coord, material, normal, pointId):
        # isotropic etch rate
        return -1.0

    def getVectorVelocity(self, coord, material, normal, pointId):
        return [0.0, 0.0, 0.0]

def main():
    # Simulation parameters
    extent = 30.0
    gridDelta = 1.0
    
    # Bounds and boundary conditions
    bounds = [-extent, extent, -extent, extent, -extent, extent]
    boundaryCons = [
        viennals.BoundaryConditionEnum.REFLECTIVE_BOUNDARY,
        viennals.BoundaryConditionEnum.REFLECTIVE_BOUNDARY,
        viennals.BoundaryConditionEnum.INFINITE_BOUNDARY
    ]
    
    # Create substrate domain
    substrate = viennals.Domain(bounds, boundaryCons, gridDelta)

    
    # Create initial plane geometry
    origin = [0.0, 0.0, 0.0]
    planeNormal = [0.0, 0.0, 1.0]
    plane = viennals.Plane(origin, planeNormal)
    viennals.MakeGeometry(substrate, plane).apply()
    
    # Create spheres used for booling
    print("Creating spheres...")
    sphere = viennals.Domain(bounds, boundaryCons, gridDelta)
    
    # Define boolean operation (Relative Complement: substrate - sphere)
    boolOp = viennals.BooleanOperation(
        substrate, sphere, viennals.BooleanOperationEnum.RELATIVE_COMPLEMENT
    )
    
    # Sphere 1
    origin = [-12.0, -5.0, -15.0]
    radius = 10.0
    viennals.MakeGeometry(sphere, viennals.Sphere(origin, radius)).apply()
    boolOp.apply()
    
    # Sphere 2
    origin = [-7.0, -30.0, -20.0]
    radius = 8.0
    viennals.MakeGeometry(sphere, viennals.Sphere(origin, radius)).apply()
    boolOp.apply()
    
    # Sphere 3
    origin = [5.0, 15.0, -2.0]
    radius = 8.0
    viennals.MakeGeometry(sphere, viennals.Sphere(origin, radius)).apply()
    boolOp.apply()
    
    # Sphere 4
    origin = [2.0, 8.0, -27.0]
    radius = 8.0
    viennals.MakeGeometry(sphere, viennals.Sphere(origin, radius)).apply()
    boolOp.apply()
    
    # Now etch the substrate isotropically
    velocities = VelocityField()
    
    print("Advecting")
    
    advectionKernel = viennals.Advect()
    advectionKernel.insertNextLevelSet(substrate)
    advectionKernel.setVelocityField(velocities)
    advectionKernel.setIgnoreVoids(True)
    
    passedTime = 0.0
    numberOfSteps = 50
    
    for i in range(numberOfSteps):
        print(f"\rAdvection step {i} / {numberOfSteps}", end="", flush=True)
        
        mesh = viennals.Mesh()
        viennals.ToSurfaceMesh(substrate, mesh).apply()
        viennals.VTKWriter(mesh, f"void-{i}.vtp").apply()
        
        advectionKernel.apply()
        passedTime += advectionKernel.getAdvectedTime()
        
    print()
    print(f"Time passed during advection: {passedTime}")

if __name__ == "__main__":
    main()
